### PR TITLE
ENH: provide Balancer with rng= argument, provide get_rng helper

### PR DESCRIPTION
**Pull request by yarikoptic**
Or should we name it 'random_state' to more inline with scikit.learn and actual beast it returns, @hanke -- what you think?

Closes: #492

### CODE REVIEW

**Comment by effigies** [comment]
I don't think you want this as local to `Balancer._call`, as then you'll always get the same balance, even when `Balancer(count=2)`. Setting in `generate` before the `for i in xrange...`

To make sure we don't over-determine, I think there needs to be a test to the effect of:

```Python
b = Balancer(count=3, rng=XXX)
balanced = list(b.generate(ds))
assert not np.array_equal(balanced[0], balanced[1])
assert not np.array_equal(balanced[0], balanced[2])
assert not np.array_equal(balanced[1], balanced[2])
```

Really, what would be ideal is to have a multi-processing test, to make sure that threads/processes aren't interfering with each other's state, but that's hard. For my [own version](https://github.com/PyMVPA/PyMVPA/compare/master...effigies:balancer_local_determinism), I ran a searchlight while logging excluded trials, and made sure they repeated. But, I was only doing it for `count=1` (my use case).

I don't have any opinions about names.

**Comment by yarikoptic** [comment]
Thanks @effigies -- all this `generate`-vs-`call` is somewhat a fragile point here indeed :-/  I think I have done it correctly now.  But then I have realized that probably the same logic should be within AttributePermutator, BUT somehow if I comment out that instantiation of `rng` within its `generate`, bloody thing still passes the test which supposed to check that all permutations are different... so either a test is incorrect, or a I am missing the point.

Also as travis points out `numpy.random.choice` is apparently a new addition :-/  so I guess I would need to RF it to use `np.random.permutation()[:count]` if we decide to stick with using `numpy` here instead of as originally pure Python's `random`.

File: mvpa2/tests/test_generators.py [hunk 1]
```diff
@@ -221,6 +241,24 @@ def test_balancer():
...
+    bal = Balancer(apply_selection=True, count=5, rng=1)
+    assert_false(np.any(bal(ds).sa.ids != bal(ds).sa.ids))
+
+    # But results should differ if we use .generate to produce those multiple
+    # balanced datasets
+    b = Balancer(apply_selection=True, count=3, rng=1)
+    balanced = list(b.generate(ds))
+    assert_false(all(balanced[0].sa.ids == balanced[1].sa.ids))
+    assert_false(all(balanced[0].sa.ids == balanced[2].sa.ids))
+    assert_false(all(balanced[1].sa.ids == balanced[2].sa.ids))
```
Comment (effigies) [inline mvpa2/tests/test_generators.py]: In addition to the sequence not being constant, I would check that the sequence is repeated on a new `generate` call:

```Python
assert_true(all(np.array_equal(ds_a.sa.ids, ds_b.sa.ids)
            for ds_a, ds_b in zip(balanced, b.generate(ds)))
```

**Comment by effigies** [comment]
So the options for `Balancer` and `AttributePermutator` are:

a) `rng=None` => use global random state
b) `rng=RandomState()` => local random state, updates with each call
c) `rng=int()` => local, determined random state, resets with each call

First, it's not clear to me what the use case for (b) is, except, I suppose, it lets you seed your own at the beginning and get deterministic output without repetitive behavior.

My use case would almost always be:

```Python
Balancer(..., rng=np.random.randint(2**32))
```

Basically, I want a constant seed throughout the life of the object, but I don't want to hard-code one, and I really don't care what its value is, except perhaps to log it.

Any chance of making this canonical and adding:

d) `rng='nonce'` => local, randomly determined initial state, resets with each call

```Python
self._rng = np.random.randint(2**32) if rng == 'nonce' else rng
```

Feel free to reject if this is getting too custom. :-) Also, the name `nonce` isn't important, if you have a better.

File: mvpa2/tests/test_generators.py (hunk 1 above)
Comment (yarikoptic) [inline mvpa2/tests/test_generators.py]: done, pushed. THANKS!

**Comment by yarikoptic** [comment]
yes, use case for b) is quite scarce.  I guess it would be a localized version of mvpa2.seed which resets both np.random and random RNGs at that point, so that entire analysis could be reproduced.  Here it could be used if someone doesn't need reproducible .generate runs, but overall wants Balancer or AttributePermutator produce reproducible results through a sequence of calls.

`nonce` (or worded alike) would be a neat convenience I would say.  But not sure if I would like to jump implementing it right away... It feels that we would like overall overhaul and unification of such per-instance seedings, and if so, we should unify handling of string values passed as well then... would you be kind to file a separate `[wishlist]` issue for that?

**Comment by effigies** [comment]
Sounds good. Will post shortly.

Final review:

- Tests: Look good.
- Docs: I would add a description of the `rng` parameter to the `AttributePermutator` docstring. And it would be good to describe the separate impacts of a constant seed on `__call__` and `generate`.
- Submitted an optional patch which clears up some execution-ordering edge cases.

**Comment by yarikoptic** [comment]
@hanke , @effigies -- what do you think about this PR in current state and 2d82c6c RFing in particular ;)

File: mvpa2/generators/resampling.py [hunk 2]
```diff
@@ -76,31 +76,38 @@ def __init__(self,
...
+    def _get_call_kwargs(self, ds):
+        attr, collection = ds.get_attr(self._attr)
+        # _call might need to operate on the dedicated instantiated rng
+        # e.g. if seed int is provided
+        return {
+            'limit_filter': get_limit_filter(self._limit, collection),
+            'rng': get_rng(self._rng)
+        }
+
+    def _call(self, ds, limit_filter=None, rng=None):
```
Comment (effigies) [inline mvpa2/generators/resampling.py]: If I'm reading this correctly, `_call` must *always* be called with `rng` (or L150) will fail.

Does it make sense for these to default to `None`, then?

**Comment by effigies** [comment]
This is much better than hidden state. If it passes, then I'm good with it.

File: mvpa2/generators/resampling.py (hunk 2 above)
Comment (yarikoptic) [inline mvpa2/generators/resampling.py]: Well, it is more just to make them keyword arguments... There is no way to enforce them being provided though in python 2 afaik... But they always should be provided unless someone provides wrong `_call_kwargs`... So overall I wouldn't sweat about assuring then not being None for this one ... Tests pass ;-)

**Comment by yarikoptic** [comment]
cool, tests pass, so I will just wait for @hanke 's blessing ;)

**Comment by yarikoptic** [comment]
I was even thinking to refactor more to provide "RepeatingNode" subclass of Node which would absorb common logic here for "generate" between Balancer and Permutator but since it was only 2 of them (didn't find any other on a quick look) decided to leave it alone for now at least ;)

**Comment by yarikoptic** [comment]
ok no complaints, merging, release is unavoidable! (after daily even more thorough testing is ran, so in few days)
